cmd/prometheus-config-reloader/main.go

```diff
@@ -227,3 +253,91 @@ func createOrdinalEnvvar(fromName string) error {
 	val := reg.FindString(os.Getenv(fromName))
 	return os.Setenv(statefulsetOrdinalEnvvar, val)
 }
+
+type prometheusReloadTracker struct {
+	baseURL *url.URL
+	client  http.Client
+	logger  log.Logger
+
+	lastReload  time.Time
+	lastSuccess bool
+}
+
+func (prt *prometheusReloadTracker) getConfigStatus(ctx context.Context) (bool, time.Time, error) {
+	r, err := http.NewRequestWithContext(ctx, "GET", prt.baseURL.JoinPath("/api/v1/status/runtimeinfo").String(), nil)
+	if err != nil {
+		return false, time.Time{}, err
+	}
+
+	resp, err := prt.client.Do(r)
+	if err != nil {
+		return false, time.Time{}, fmt.Errorf("%s: %w", r.URL.String(), err)
+	}
+	defer resp.Body.Close()
+
+	if resp.StatusCode/100 != 2 {
+		return false, time.Time{}, fmt.Errorf("%s: invalid status code: %d", r.URL.String(), resp.StatusCode)
+	}
+
+	var runtimeInfo = struct {
+		Status string `json:"status"`
+		Data   struct {
+			ReloadConfigSuccess bool      `json:"reloadConfigSuccess"`
+			LastConfigTime      time.Time `json:"lastConfigTime"`
+		} `json:"data"`
+	}{}
+
+	if err := json.NewDecoder(resp.Body).Decode(&runtimeInfo); err != nil {
+		return false, time.Time{}, fmt.Errorf("invalid response: %w", err)
+	}
+
+	if runtimeInfo.Status != "success" {
+		return false, time.Time{}, fmt.Errorf("unexpected status: %s", runtimeInfo.Status)
+	}
+
+	return runtimeInfo.Data.ReloadConfigSuccess, runtimeInfo.Data.LastConfigTime, nil
+}
+
+func (prt *prometheusReloadTracker) preReload(ctx context.Context) error {
+	success, last, err := prt.getConfigStatus(ctx)
+	if err != nil {
+		return fmt.Errorf("failed to get config status: %w", err)
+	}
+
+	level.Debug(prt.logger).Log("msg", "Pre reload hook", "success", success, "last_config_reload", last)
+	prt.lastReload = last
+	prt.lastSuccess = success
+
+	return nil
+}
+
+func (prt *prometheusReloadTracker) postReload(ctx context.Context) error {
+	t := time.NewTicker(time.Second)
+	defer t.Stop()
+
+	for {
+		success, last, err := prt.getConfigStatus(ctx)
+		if err != nil {
+			return fmt.Errorf("failed to get config status: %w", err)
+		}
+		level.Debug(prt.logger).Log("msg", "Post reload hook in progress", "success", success, "last_config_reload", last)
+
+		// The configuration has been updated successfully.
+		if success && last.After(prt.lastReload) {
+			level.Debug(prt.logger).Log("msg", "Post reload hook completed", "success", success, "last_config_reload", last)
+			return nil
+		}
+
+		// The previous configuration was valid but the current configuration
+		// isn't so there's no need to poll again.
+		if !success && prt.lastSuccess {
+			return fmt.Errorf("configuration reload failed")
+		}
+
+		select {
+		case <-ctx.Done():
+			return fmt.Errorf("%w: last configuration status: success=%v, success_timetamp=%v", ctx.Err(), success, last)
+		case <-t.C:
```

Comment:
Dont you need to reset the timer?

Reply:
No it will automatically resend to the channel at the next tick.